zypp/base/String.cc

```diff
@@ -388,6 +388,91 @@ namespace zypp
       return std::string( buf.begin(), buf.end() );
     }
 
+
+    std::string bEscape( std::string str_r, const C_Str & special_r )
+    {
+      if ( str_r.empty() )
+	return std::move( str_r );
+
+      if ( str_r.find_first_of( special_r ) == std::string::npos
+	&& ( ::strchr( special_r.c_str(), '\\' ) ||  !::strchr( str_r.c_str(), '\\' ) ) )
+	return std::move( str_r );
+
+      Str buf;
+      for_( s, str_r.c_str(), s+str_r.size() )
+      {
+	if ( *s == '\\' || ::strchr( special_r.c_str(), *s ) )
+	  buf << '\\';
+	buf << *s;
+      }
+      return buf;
+    }
+
+    #define RXSPECIALCHARS "\\.*+?^$[()|{"
+
+    std::string rxEscapeStr( std::string str_r )
+    {
+      return bEscape( std::move(str_r), RXSPECIALCHARS );
+    }
+
+    std::string rxEscapeGlob( std::string str_r )
+    {
+      if ( str_r.empty() )
+	return std::move( str_r );
+
+      if ( str_r.find_first_of( RXSPECIALCHARS ) == std::string::npos )
+	return std::move( str_r );
+
+      Str buf;
+      for_( s, str_r.c_str(), s+str_r.size() )
+      {
+	if ( *s == '\\' )	// + next char literally
+	{
+	  buf << '\\';
+	  if ( *(s+1) ) { ++s; buf << *s; }
+	}
+	else if ( *s == '?' )	// translate
+	{
+	  buf << '.';
+	}
+	else if ( *s == '*' )	// translate
+	{
+	  buf << ".*";
+	}
+	else if ( *s == '[' )	// character class if closing ] is found, else literally
+	{
+	  const char * e = s+1;
+	  if ( *e == '^' || *e == '!' )	// negated cclass
+	    ++e;
+	  if ( *e == ']' )		// ] in cclass
```

Comment:
Inside a character class `[]` nothing is escapable (neither in RX nor GLOG). A `\` is a `\` and a `]` is a`]`. However to prevent a `]`  from being interpreted as the character class closing `]`, it must appear as the 1st char or after the negating `^` or `!`.

Reply:
Makes sense